js/webapp.js

```diff
@@ -783,6 +783,37 @@ WebApp.prototype = {
     }
   },
 
+  resolveAllImports(pluginDefs) {
+    let unresolvedPlugins = [];
+    installLog.info(`Resolving imports for ${pluginDefs.length} remaining plugins`);
+    pluginDefs.forEach((plugin) => {
+      installLog.debug(
+        `${plugin.identifier}: ${plugin.dataServicesGrouped? 'has' : 'does not have'}`
+          + ' services')
+      const urlBase = zLuxUrl.makePluginURL(this.options.productCode, 
+                                            plugin.identifier);
+      try {
+        this._resolveImports(plugin, urlBase);
+      } catch (e) {
+        unresolvedPlugins.push(plugin);
+        //keep in array and retry
+      }
+    });
+    if (unresolvedPlugins.length === 0) {
+      installLog.info(`All imports resolved for all plugins.`);
+      return true;
+    } else if (unresolvedPlugins.length !== pluginDefs.length) {
```

Comment:
It's not immediately obvious to me why you should need to re-drive this method now that you've pushed this to the post-processing phase (i.e. after all plugins have been initially loaded). So, if it really is needed, I think a comment would be helpful.

How well is an import loop handled? Does this algorithm terminate? I think it does, but did you test? Maybe later we can worry about being able to report errors of that type in more detail. For now it's enough that this algorithm terminates, and the full list of failures is apparent in the log.

Reply:
let me retest this to see: i was concerned about imports-of-imports-of-imports even if it is not a recommended thing to do. I'll try to make the react app import the iframe app dataservice which imports the angular dataservice and see what the behavior is in this function.